libpoolprotocols/PoolURI.cpp

```diff
@@ -0,0 +1,98 @@
+#include <libpoolprotocols/PoolURI.h>
+#include <map>
+
+using namespace dev;
+using namespace std;
+
+static void toLower(string &str)
+{
+	string::iterator end = str.end();
+	for (string::iterator it = str.begin(); it != end; it++)
+		*it = tolower(*it);
+}
+
+typedef struct {
+	ProtocolFamily family;
+	SecureLevel secure;
+	unsigned version;
+} SchemeType;
+
+static map<string, SchemeType> s_schemes = {
+	{"stratum",	{ProtocolFamily::STRATUM, SecureLevel::NONE,  0}},
+	{"stratum1",	{ProtocolFamily::STRATUM, SecureLevel::NONE,  1}},
+	{"stratum2",	{ProtocolFamily::STRATUM, SecureLevel::NONE,  2}},
+	{"stratum+tcp",	{ProtocolFamily::STRATUM, SecureLevel::NONE,  0}},
+	{"stratum1+tcp",{ProtocolFamily::STRATUM, SecureLevel::NONE,  1}},
+	{"stratum2+tcp",{ProtocolFamily::STRATUM, SecureLevel::NONE,  2}},
+	{"stratum+tls",	{ProtocolFamily::STRATUM, SecureLevel::TLS12, 0}},
+	{"stratum1+tls",{ProtocolFamily::STRATUM, SecureLevel::TLS12, 1}},
+	{"stratum2+tls",{ProtocolFamily::STRATUM, SecureLevel::TLS12, 2}},
+	{"stratum+ssl",	{ProtocolFamily::STRATUM, SecureLevel::TLS12, 0}},
+	{"stratum1+ssl",{ProtocolFamily::STRATUM, SecureLevel::TLS12, 1}},
+	{"stratum2+ssl",{ProtocolFamily::STRATUM, SecureLevel::TLS12, 2}},
+	{"http",	{ProtocolFamily::GETWORK, SecureLevel::NONE,  0}},
+	{"https",	{ProtocolFamily::GETWORK, SecureLevel::TLS12, 0}},
+	{"http+tls",	{ProtocolFamily::GETWORK, SecureLevel::TLS12, 0}},
+	{"http+ssl",	{ProtocolFamily::GETWORK, SecureLevel::TLS12, 0}}
+};
+
+bool URI::KnownScheme()
+{
+	return s_schemes.find(m_scheme) != s_schemes.end();
+}
+
+ProtocolFamily URI::ProtoFamily()
+{
+	return s_schemes[m_scheme].family;
+}
+
+unsigned URI::ProtoVersion()
+{
+	return s_schemes[m_scheme].version;
+}
+
+SecureLevel URI::ProtoSecureLevel()
+{
+	return s_schemes[m_scheme].secure;
+}
+
+string URI::KnownSchemes(ProtocolFamily family)
+{
+	string schemes;
+	for(const auto&s : s_schemes)
+		if (s.second.family == family)
+			schemes += s.first + " ";
+	return schemes;
+}
+
+void URI::parse(const string &uri)
```

Comment:
user + password is also part of the URI specification: https://tools.ietf.org/html/rfc3986#section-3.2

Reply:
Also i just googled and found this: https://github.com/bnoordhuis/uriparser2
Not saying to use THIS code but i suggest using an existing parser instead of writing our own.

Reply:
maybe CppNetlibUri included in hunter,  not sure if it works on windows. i tried CppNetlib  before and the hunter package had problems

Reply:
Yeah, I looked for something hunter supported and found CppNetlibUri... Turns out it's a URI builder, not parser! Uriparser2 looks good but I've no clue how to pull it down and integrate it to the build. Hence the current code. I'd prefer not to have to learn hunter.

I'm also thinking of the @ syntax for user spec. I'm leery of the user:password notation as it has been deprecated long ago, and it's considered really bad manners to require passwords in the clear on the command line. Folks are concerned with BGP attacks but hard code passwords into their scripts... right! If we are serious about security then passwords are either prompted for, or included in some access protected file.

Reply:
For passwords, I'll accept the the user:password notation, but implement a special case where if the password is 'prompt' we prompt for it.

Reply:
Trying to make a cmakefile for uriparser2, but it does not compile on msvc at the moment.
Still trying.